### ISSUE TITLE
Questions about UDP relay mode quic

### ISSUE THREAD

**Issue by Blankwonder** [tuic-protocol/tuic#97]
Hi. I have already implemented the `native` UDP relay mode of TUIC. But since the default max datagram payload size is too small for QUIC, the native mode may encounter many issues.

So I tried to implement the `quic` mode. But I'm not sure why the client should use a unidirectional stream instead of a bidirectional stream. How the client gets the packets from the server? Does the server always send the UDP datagram back via a QUIC datagram frame? Or the server will initiate another unidirectional stream?

Thanks.

**Comment by zephyrchien** [comment]
In `quic` mode, the server will open a **new** `uni stream` to send each one packet back, and then close this stream, like sending a stateless datagram. So there is no much difference between `native` and `quic` mode from an implementation perspective.

**Comment by Blankwonder** [comment]
> In `quic` mode, the server will open a **new** `uni stream` to send each one packet back, and then close this stream, like sending a stateless datagram. So there is no much difference between `native` and `quic` mode from an implementation perspective.

This doesn't seem necessary. A new stream for every packet may create additional QUIC frames and other unnecessary costs to maintain the stream until all data frames are acked. A bidirectional stream seems more reasonable.

**Comment by zephyrchien** [comment]
Yes, there's room for improvement in the protocol. I assume it was for convenience, where facilities(`Associate`, `Dissociate`) can be reused, since opening a new stream will not introduce extra RTT.

**Comment by Blankwonder** [comment]
OK, I got the idea of why to use a new stream for every packet. And I just completed the support of the `quic` mode. 

But I found another issue. The TUIC server uses the default parameters to create a QUIC session. The maximum of concurrency uni streams allowed is 100, which means the client can only send 100 UDP packets without blocking. It's obvious not enough.

Did I miss something, or should the server program choose a much more higher value of `initial_max_streams_uni`?

**Comment by EAimTY** [comment]
Using a unidirection stream for every packet in `quic` mode is for avoiding several concurrent packets from blocking each other, and for avoiding the packet loss.

A new packet fragment mechanism for `native` mode is under developing in the `refactor` branch. It will be out in the next major version of the protocol.

**Comment by Blankwonder** [comment]
@EAimTY Please consider adding a new mode using a single stream to relay all UDP packets. It works just like UDP over TCP and should be fine for most cases.

While if using a separated uni-direction stream for every packet, 100Mbps bandwidth requires about 10000+ streams. I don't think QUIC protocol or its implementations are ready for such a large number of stream concurrency.
